README.md
<p float="left">
   <img src="https://fastmachinelearning.github.io/hls4ml/img/logo.jpg" alt="hls4ml" width="400"/>
</p>

[![DOI](https://zenodo.org/badge/108329371.svg)](https://zenodo.org/badge/latestdoi/108329371)
[![PyPI version](https://badge.fury.io/py/hls4ml.svg)](https://badge.fury.io/py/hls4ml)
[![Supported Python versions](https://img.shields.io/pypi/pyversions/hls4ml.svg)](https://pypi.org/project/hls4ml/)

A package for machine learning inference in FPGAs. We create firmware implementations of machine learning algorithms using high level synthesis language (HLS). We translate traditional open-source machine learning package models into HLS that can be configured for your use-case!

If you have any questions, comments, or ideas regarding hls4ml or just want to show us how you use hls4ml, don't hesitate to reach us through the [discussions](https://github.com/fastmachinelearning/hls4ml/discussions) tab.

# Documentation & Tutorial

For more information visit the webpage: [https://fastmachinelearning.org/hls4ml/](https://fastmachinelearning.org/hls4ml/)

Detailed tutorials on how to use `hls4ml`'s various functionalities can be found [here](https://github.com/hls-fpga-machine-learning/hls4ml-tutorial).

# Installation
```
pip install hls4ml
```

To install the extra dependencies for profiling:

```
pip install hls4ml[profiling]
```

# Getting Started
### Creating an HLS project
```Python
import hls4ml

#Fetch a keras model from our example repository
#This will download our example model to your working directory and return an example configuration file
config = hls4ml.utils.fetch_example_model('KERAS_3layer.json')

print(config) #You can print the configuration to see some default parameters

#Convert it to a hls project
hls_model = hls4ml.converters.keras_to_hls(config)

# Print full list of example models if you want to explore more
hls4ml.utils.fetch_example_list()
```

### Building a project with Xilinx Vivado HLS (after downloading and installing from [here](https://www.xilinx.com/products/design-tools/vivado/integration/esl-design.html))
Note: Vitis HLS is not yet supported. Vivado HLS versions between 2018.2 and 2020.1 are recommended.

```Python
#Use Vivado HLS to synthesize the model
#This might take several minutes
hls_model.build()

#Print out the report if you want
hls4ml.report.read_vivado_report('my-hls-test')
```

# Citation
If you use this software in a publication, please cite the software
```bibtex
@software{vloncar_2021_5680908,
  author       = {{FastML Team}},
  title        = {fastmachinelearning/hls4ml},
  year         = 2021,
  publisher    = {Zenodo},
  doi          = {10.5281/zenodo.1201549},
  url          = {https://github.com/fastmachinelearning/hls4ml}
}
```
and first publication:
```bibtex
@article{Duarte:2018ite,
    author = "Duarte, Javier and others",
    title = "{Fast inference of deep neural networks in FPGAs for particle physics}",
    eprint = "1804.06913",
    archivePrefix = "arXiv",
    primaryClass = "physics.ins-det",
    reportNumber = "FERMILAB-PUB-18-089-E",
    doi = "10.1088/1748-0221/13/07/P07027",
    journal = "JINST",
    volume = "13",
    number = "07",
    pages = "P07027",
    year = "2018"
}
```
Additionally, if you use specific features developed in later papers, please cite those as well. For example, CNNs:
```bibtex
@article{Aarrestad:2021zos,
    author = "Aarrestad, Thea and others",
    title = "{Fast convolutional neural networks on FPGAs with hls4ml}",
    eprint = "2101.05108",
    archivePrefix = "arXiv",
    primaryClass = "cs.LG",
    reportNumber = "FERMILAB-PUB-21-130-SCD",
    doi = "10.1088/2632-2153/ac0ea1",
    journal = "Mach. Learn. Sci. Tech.",
    volume = "2",
    number = "4",
    pages = "045015",
    year = "2021"
}
@article{Ghielmetti:2022ndm,
    author = "Ghielmetti, Nicol\`{o} and others",
    title = "{Real-time semantic segmentation on FPGAs for autonomous vehicles with hls4ml}",
    eprint = "2205.07690",
    archivePrefix = "arXiv",
    primaryClass = "cs.CV",
    reportNumber = "FERMILAB-PUB-22-435-PPD",
    doi = "10.1088/2632-2153/ac9cb5",
    journal ="Mach. Learn. Sci. Tech.",
    year = "2022"
}
```
binary/ternary networks:
```bibtex
@article{Loncar:2020hqp,
    author = "Ngadiuba, Jennifer and others",
    title = "{Compressing deep neural networks on FPGAs to binary and ternary precision with HLS4ML}",
    eprint = "2003.06308",
    archivePrefix = "arXiv",
    primaryClass = "cs.LG",
    reportNumber = "FERMILAB-PUB-20-167-PPD-SCD",
    doi = "10.1088/2632-2153/aba042",
    journal = "Mach. Learn. Sci. Tech.",
    volume = "2",
    pages = "015001",
    year = "2021"
}
```


docs/release_notes.rst
========================
Release Notes
========================

See `here <https://github.com/fastmachinelearning/hls4ml/releases>`__ for official releases on Github.

----

**v0.6.0 / coris**

What's changed:

* ``VivadoAccelerator`` backend: target ``pynq-z2`` and ``zcu102`` boards directly from hls4ml by @nicologhielmetti
* Updated ``PyTorch`` and ``ONNX`` converters by @Duchstf
* ``line_buffer`` Conv2D implementation for ``io_stream``: reduced resource usage and latency by @Keb-L, @violatingcp, @vloncar
* Support ``QConv2DBatchnorm`` layer from ``QKeras`` by @nicologhielmetti
* Improved profiling plots - easier to compare original vs ``hls4ml`` converted models by @maksgraczyk
* Better derivation of data types for ``QKeras`` models by @jmduarte, @thesps
* Improved CI by @thesps
* More support for models with branches, skip connections, ``Merge`` and ``Concatenate`` layers by @jmduarte, @vloncar
* Support for ``Dense`` layers over multi-dimensional tensors by @vloncar
* Overall improvements by @vloncar, @jmduarte, @thesps, @jmitrevs & others

New contributors:

* @siorpaes made their first contribution in https://github.com/fastmachinelearning/hls4ml/pull/424
* @jmitrevs made their first contribution in https://github.com/fastmachinelearning/hls4ml/pull/403
* @anders-wind made their first contribution in https://github.com/fastmachinelearning/hls4ml/pull/302
* @KOVI89alipes made their first contribution in https://github.com/fastmachinelearning/hls4ml/pull/318
* @maksgraczyk made their first contribution in https://github.com/fastmachinelearning/hls4ml/pull/323
* @Keb-L made their first contribution in https://github.com/fastmachinelearning/hls4ml/pull/332
* @ConsVin made their first contribution in https://github.com/fastmachinelearning/hls4ml/pull/307
* @nicologhielmetti made their first contribution in https://github.com/fastmachinelearning/hls4ml/pull/298

**Full Changelog**: https://github.com/fastmachinelearning/hls4ml/compare/v0.5.0...v0.6.0

----

**v0.5.0 / bartsia**

What's new:

* Streaming IO layer implementations, especially of Convolutional layers, accessed through the config with `IOType: io_stream`. Scales CNN support to much larger models than previously possible (see `arXiv:2101.05108 <https://arxiv.org/abs/2101.05108>`__)
* New `documentation and API reference <https://fastmachinelearning.org/hls4ml/>`__
* Further optimizations for QKeras / quantization aware training. A 'shift' operation is now used for `po2` quantizers
* Allow redefinition of weights directory for standalone project compilation
* ``profiling`` for PyTorch models

Deprecated:

* ``IOType : io_serial`` is deprecated, and superceded by new ``IOType: io_stream``

Bugfixes:

* Fix to Initiation Interval and different min/max latency for ``Strategy: Resource``
* Fix warnings in ``hls4ml`` command line script flow
* Write yml config from Python API - for mixed API / command line flow

----

**v0.4.0 / aster**

What's new:

* Support for GarNet layer (see `arXiv:2008.03601 <https://arxiv.org/abs/2008.03601>`__)
* Input layer precision added to config generator utility
* New 'SkipOptimizers' config option. Now you can run all Optimizers by default (as in v0.3.0) but subtract any specified by ``SkipOptimizer`` e.g. ``hls_config['SkipOptimizers'] = ['fuse_consecutive_batch_normalization']``
* Print out the latency report from Cosimulation

Bugfixes:

* Fixes related to tensorflow 2.3: new Functional API, changes to handling of Input layer
* Fix error with config generator utility and activation layers gor ``granularity='name'``
* Fix issue with reloading of emulation library after configuration change
* Fix to handling of layers with ``use_bias=False`` and merged Dense and BatchNormalization

----

**v0.3.0**


* Installing from ``PyPI``
* Create configuration dictionary from model object
* Run 'C Simulation' from Python with ``hls_model.predict(X)``
* Trace model layer output with ``hls_model.trace(X)``
* Write HLS project, run synthesis flow from Python
* QKeras support: convert models trained using layers and quantizers from QKeras
* Example models moved to separate repo, added API to retrieve them
* New Softmax implementations
* Minor fixes: weights exported at higher precision, concatenate layer shape corrected

----

**v0.2.0:**


* ``tf_to_hls`` tool for converting tensorflow models (protobufs ``.pb``\ )
* Support for larger ``Conv1D/2D`` layers
* Support for binary and ternary layers from `QKeras <https://github.com/google/qkeras>`_.
* API enhancements (custom layers, multiple backends)
* :doc:`Profiling <api/profiling>` support
* ``hls4ml report``\ command to gather HLS build reports, ``hls4ml build -l`` for Logic Synthesis
* Support for all-in-one Keras's ``.h5`` files (obtained with Keras's ``save()`` function, without the need for separate ``.json`` and ``.h5`` weight file).
* Fused Batch Normalisation into Dense layer optimsation.

----

**v0.1.6:**


* Support for larger Dense layers (enabled with Strategy: Resource in the configuration file)
* Binary/Ternary NN refinements
* Built-in optimization framework
* Optional C/RTL validation

----

**v0.1.5**\ : Per-layer precision and reuse factor

----

**v0.1.3**\ : Adding PyTorch support

----

**v0.1.2**\ : First beta release


* some bug fixes for pipelining and support for layer types

----

**v0.0.2**\ : first alpha release


* full translation of DNNs from Keras
* an example Conv1D exists
* parallel mode is supported (serial mode, not yet)


hls4ml/utils/config.py
import json

import hls4ml


def create_config(output_dir='my-hls-test', project_name='myproject', backend='Vivado', **kwargs):

    backend_list = hls4ml.backends.get_available_backends()
    if backend.lower() not in backend_list:
        raise Exception(f'Unknown backend: {backend}')

    backend = hls4ml.backends.get_backend(backend)

    backend_config = backend.create_initial_config(**kwargs)

    config = {}
    config['OutputDir'] = output_dir
    config['ProjectName'] = project_name
    config['Backend'] = backend.name
    config.update(backend_config)

    return config


def _get_precision_from_quantizer(quantizer):
    import qkeras

    if isinstance(quantizer, str):
        quantizer_obj = qkeras.get_quantizer(quantizer)
        quantizer = {}
        # Some activations are classes with get_config method
        if hasattr(quantizer_obj, 'get_config'):
            quantizer['class_name'] = quantizer_obj.__class__.__name__
            quantizer['config'] = quantizer_obj.get_config()
        # Some activations are just functions
        else:
            quantizer['class_name'] = quantizer_obj.__name__

    supported_quantizers = ['quantized_bits', 'quantized_relu', 'quantized_tanh', 'quantized_po2', 'quantized_relu_po2']
    signed = True
    if quantizer['class_name'] in supported_quantizers:
        bits = int(quantizer['config']['bits'])
        # if integer isn't specified, it should be the same as bits
        integer = int(quantizer['config'].get('integer', bits - 1)) + 1
        if quantizer['class_name'] == 'quantized_relu':
            signed = False
            integer -= 1
    elif quantizer['class_name'] in ['binary', 'stochastic_binary', 'binary_tanh']:
        bits = 2
        integer = 2

    elif quantizer['class_name'] in ['ternary', 'stochastic_ternary', 'ternary_tanh']:
        bits = 2
        integer = 2
    else:
        raise Exception('ERROR: Unsupported quantizer: {}'.format(quantizer['class_name']))

    decimal = bits - integer
    signed = '' if signed else 'u'
    if decimal > 0:
        return f'ap_{signed}fixed<{bits},{integer}>'
    else:
        return f'ap_{signed}int<{bits}>'


def config_from_keras_model(model, granularity='model', default_precision='ap_fixed<16,6>', default_reuse_factor=1):
    """Create an HLS conversion config given the Keras model.

    This function serves as the initial step in creating the custom conversion configuration.
    Users are advised to inspect the returned object to tweak the conversion configuration.
    The return object can be passed as `hls_config` parameter to `convert_from_keras_model`.

    Args:
        model: Keras model
        granularity (str, optional): Granularity of the created config. Defaults to 'model'.
            Can be set to 'model', 'type' and 'layer'.

            Granularity can be used to generate a more verbose config that can be fine-tuned.
            The default granulrity ('model') will generate config keys that apply to the whole
            model, so changes to the keys will affect the entire model. 'type' granularity will
            generate config keys that affect all layers of a given type, while the 'name' granularity
            will generate config keys for every layer separately, allowing for highly specific
            configuration tweaks.
        default_precision (str, optional): Default precision to use. Defaults to 'ap_fixed<16,6>'.
        default_reuse_factor (int, optional): Default reuse factor. Defaults to 1.

    Raises:
        Exception: If Keras model has layers not supported by hls4ml.

    Returns:
        [dict]: The created config.
    """
    if granularity.lower() not in ['model', 'type', 'name']:
        raise Exception(
            f'Invalid configuration granularity specified, expected "model", "type" or "name" got "{granularity}"'
        )

    # This is a list of dictionaries to hold all the layer info we need to generate HLS
    layer_list = []

    if isinstance(model, dict):
        model_arch = model
    else:
        model_arch = json.loads(model.to_json())

    # Define supported layers
    core_layers = ['InputLayer', 'Dropout', 'Flatten', 'Reshape', 'Permute', 'Embedding']
    dense_layers = ['Dense', 'BinaryDense', 'TernaryDense']
    conv_layers = ['Conv1D', 'Conv2D', 'BinaryConv2D', 'SeparableConv2D']
    pooling_layers = [
        'MaxPooling1D',
        'MaxPooling2D',
        'GlobalMaxPooling1D',
        'GlobalMaxPooling2D',
        'AveragePooling1D',
        'AveragePooling2D',
        'GlobalAveragePooling1D',
        'GlobalAveragePooling2D',
    ]
    norm_layers = ['BatchNormalization']
    activation_layers = ['Activation', 'LeakyReLU', 'ThresholdedReLU', 'ELU', 'PReLU', 'Softmax', 'ReLU']
    merge_layers = ['Add', 'Subtract', 'Multiply', 'Average', 'Maximum', 'Minimum', 'Concatenate', 'Dot']
    qkeras_layers = ['QDense', 'QActivation', 'QConv1D', 'QConv2D', 'QBatchNormalization', 'QConv2DBatchnorm']
    upsampling_layers = ['UpSampling1D', 'UpSampling2D']
    reshaping_layers = ['ZeroPadding1D', 'ZeroPadding2D']
    graph_layers = ['GarNet', 'GarNetStack']
    rnn_layers = ['SimpleRNN', 'LSTM', 'GRU']
    # Define layers to skip because they're not configurable or not converted to HLS
    skip_layers = ['Dropout', 'Flatten', 'Reshape', 'Permute']
    # All supported layers
    supported_layers = (
        core_layers
        + dense_layers
        + conv_layers
        + pooling_layers
        + norm_layers
        + activation_layers
        + merge_layers
        + qkeras_layers
        + upsampling_layers
        + reshaping_layers
        + graph_layers
        + rnn_layers
        + skip_layers
    )

    keras_layer_config = None
    if model_arch['class_name'] == 'Sequential':
        print('Interpreting Sequential')
        keras_layer_config = model_arch['config']
        if 'layers' in keras_layer_config:  # Newer Keras versions have 'layers' in 'config' key
            keras_layer_config = keras_layer_config['layers']
        # Sequential doesn't have InputLayer in TF < 2.3 (Keras 2.4.0)
        if keras_layer_config[0]['class_name'] != 'InputLayer':
            input_layer = {}
            input_layer['name'] = 'input1'
            input_layer['class_name'] = 'Input'
            layer_list.append(input_layer)
    elif model_arch['class_name'] in ['Model', 'Functional']:
        print('Interpreting Model')
        keras_layer_config = model_arch['config']['layers']

    print('Topology:')
    for keras_layer in keras_layer_config:
        if keras_layer['class_name'] not in supported_layers:
            raise Exception('ERROR: Unsupported layer type: {}'.format(keras_layer['class_name']))
        if keras_layer['class_name'] in skip_layers:
            continue

        # Dictionary to fill in and append to layer_list
        layer = {}

        # Extract name for finding weights and biases
        layer['name'] = keras_layer['config']['name']
        layer['class_name'] = keras_layer['class_name']
        layer['config'] = keras_layer['config']

        if layer['class_name'] == 'InputLayer':
            layer['class_name'] = 'Input'

        if layer['class_name'] in qkeras_layers:
            layer['precision'] = {}
            for qname, qclass in layer['config'].items():
                if 'quantizer' in qname.lower():
                    pname = qname.split('_quantizer')[0]
                    if pname == 'kernel':
                        pname = 'weight'
                    if qclass is not None:
                        precision = _get_precision_from_quantizer(qclass)
                        layer['precision'][pname] = precision
                elif qname == 'activation' and layer['class_name'] == 'QActivation':
                    precision = _get_precision_from_quantizer(qclass)
                    layer['precision']['result'] = precision

        if layer['class_name'] in graph_layers:
            # Graph layer config needs access to number of input vertices from the shape of the input tensor
            # but to really compute it we'd have to track the full tensor flow as done in hls4ml.converters.keras_to_hls.
            # So here we just assume that the first input layer of the model has shape [batch_size, n_vertices, n_features]
            try:
                first_input_layer = next(kl for kl in keras_layer_config if kl['class_name'] == 'InputLayer')
                layer['n_vertices'] = first_input_layer['config']['batch_input_shape'][1]
            except:
                print('  Generating config for keras layer {}: could not estimate n_vertices. Defaulting to 128.')
                layer['n_vertices'] = 128

        print('Layer name: {}, layer type: {}'.format(layer['name'], layer['class_name']))
        layer_list.append(layer)
        if 'activation' in layer['config'] and layer['class_name'] not in activation_layers:
            act_layer = {}
            act_details = layer['config']['activation']
            if isinstance(act_details, dict):
                precision = _get_precision_from_quantizer(act_details)
                act_details = act_details['class_name']
                act_layer['precision'] = {}
                act_layer['precision']['result'] = precision
                act_layer['class_name'] = 'QActivation'
            else:
                act_layer['class_name'] = 'Activation'
            act_layer['name'] = layer['name'] + '_' + act_details
            print('  -> Activation ({}), layer name: {}'.format(act_details, layer['name']))
            layer_list.append(act_layer)

    def make_layer_config(layer):
        layer_config = {}
        if layer['class_name'] in dense_layers + conv_layers + rnn_layers:
            layer_config['Precision'] = {}
            layer_config['Precision']['weight'] = default_precision
            layer_config['Precision']['bias'] = default_precision
            layer_config['Precision']['result'] = default_precision
            layer_config['ReuseFactor'] = default_reuse_factor
            if layer['class_name'] in rnn_layers:
                layer_config['Precision']['recurrent_weight'] = default_precision

        elif layer['class_name'] in activation_layers:
            layer_config['Precision'] = default_precision
            layer_config['ReuseFactor'] = default_reuse_factor
            layer_config['table_size'] = 1024
            is_softmax = layer['class_name'] == 'Softmax'
            if 'config' in layer.keys():
                if 'activation' in layer['config'].keys():
                    is_softmax = is_softmax or (layer['config']['activation'] == 'softmax')
            if is_softmax:
                layer_config['exp_table_t'] = 'ap_fixed<18,8,AP_RND,AP_SAT>'
                layer_config['inv_table_t'] = 'ap_fixed<18,8,AP_RND,AP_SAT>'
            else:
                layer_config['table_t'] = 'ap_fixed<18,8>'

        elif layer['class_name'] in norm_layers:
            layer_config['Precision'] = {}
            layer_config['Precision']['scale'] = default_precision
            layer_config['Precision']['bias'] = default_precision
            layer_config['ReuseFactor'] = default_reuse_factor

        elif layer['class_name'] in qkeras_layers:
            if 'precision' in layer:
                layer_config['Precision'] = {}
                for name, precision in layer['precision'].items():
                    layer_config['Precision'][name] = precision
            else:
                print(
                    'WARNING: Found no precision information in QKeras layer {} ({})'.format(
                        layer['name'], layer['class_name']
                    )
                )
                layer_config['Precision'] = default_precision
            layer_config['ReuseFactor'] = default_reuse_factor

        elif layer['class_name'] in ['GarNet', 'GarNetStack']:
            ## Following code copy-pasted from hls4ml.model.hls_layers - can we factor out commonalities between the two modules?

            ## Define default precisions for various internal arrays (can be overridden from the config file)
            import math

            log2_reuse = int(math.log(default_reuse_factor, 2.0))
            n_vertices_width = int(math.log(layer['n_vertices'], 2.0))

            # We always give 10 digits for the subintegral part
            fwidth = 10
            # Integral precision for aggr_t depends on how large the temporary sum for weighed feature mean will be
            aggr_intw = max(log2_reuse, n_vertices_width - log2_reuse) + 3  # safety factor 2**3
            aggr_w = aggr_intw + fwidth
            # edge_weight_aggr_t does not need the safety factor
            ew_aggr_intw = aggr_intw - 3
            ew_aggr_w = ew_aggr_intw + fwidth

            layer_config['Precision'] = {}
            layer_config['Precision']['edge_weight'] = 'ap_ufixed<10,0,AP_TRN,AP_SAT>'
            layer_config['Precision']['edge_weight_aggr'] = f'ap_ufixed<{ew_aggr_w},{ew_aggr_intw},AP_TRN,AP_SAT>'
            layer_config['Precision']['aggr'] = f'ap_fixed<{aggr_w},{aggr_intw},AP_TRN,AP_SAT>'
            layer_config['Precision']['norm'] = 'ap_ufixed<14,4,AP_TRN,AP_SAT>'

            layer_config['ReuseFactor'] = default_reuse_factor

        elif layer['class_name'] == 'Input':
            layer_config['Precision'] = {}

            dtype = layer['config']['dtype']
            if dtype.startswith('int') or dtype.startswith('uint'):
                typename = dtype[: dtype.index('int') + 3]
                width = int(dtype[dtype.index('int') + 3 :])
                layer_config['Precision']['result'] = f'ap_{typename}<{width}>'
            # elif bool, q[u]int, ...
            else:
                layer_config['Precision']['result'] = default_precision

        else:
            layer_config['Precision'] = default_precision

        return layer_config

    config = {}

    model_config = {}
    model_config['Precision'] = default_precision
    model_config['ReuseFactor'] = default_reuse_factor
    model_config['Strategy'] = 'Latency'
    # model_config['Compression'] = False
    # model_config['Trace'] = False

    config['Model'] = model_config

    if granularity.lower() == 'type':
        type_config = {}
        for layer in layer_list:
            if layer['class_name'] in type_config:
                continue
            layer_config = make_layer_config(layer)
            type_config[layer['class_name']] = layer_config

        config['LayerType'] = type_config

    elif granularity.lower() == 'name':
        name_config = {}
        for layer in layer_list:
            layer_config = make_layer_config(layer)
            name_config[layer['name']] = layer_config

        config['LayerName'] = name_config

    return config


def config_from_pytorch_model(model, granularity='model', default_precision='ap_fixed<16,6>', default_reuse_factor=1):
    """Generate configuration dictionary from a Pytorch model.

    Parameters
    ----------
    model : Pytorch model object.
        Model to be converted to hls model object.
    granularity : string, optional
        How granular you want the configuration to be.
    default_precision : string, optional
        Defines the precsion of your inputs, outputs, weights and biases.
        It is denoted by ap_fixed<X,Y>, where Y is the number of bits representing
        the signed number above the binary point (i.e. the integer part),
        and X is the total number of bits. Additionally, integers in fixed precision
        data type (ap_int<N>, where N is a bit-size from 1 to 1024) can also be used.
    default_reuse_factor : int, optional
        Reuse factor for hls model

    Returns
    -------
    config : dict
        configuration dictionary to be used in Pytorch converter.

    See Also
    --------
    hls4ml.config_from_keras_model, hls4ml.convert_from_onnx_model

    Examples
    --------
    >>> import hls4ml
    >>> config = hls4ml.utils.config_from_keras_model(model, granularity='model')
    >>> hls_model = hls4ml.converters.convert_from_keras_model(model, hls_config=config)

    """

    config = {}

    model_config = {}
    model_config['Precision'] = default_precision
    model_config['ReuseFactor'] = default_reuse_factor
    model_config['Strategy'] = 'Latency'

    config['Model'] = model_config

    return config


def config_from_onnx_model(model, granularity='model', default_precision='ap_fixed<16,6>', default_reuse_factor=1):
    """Generate configuration dictionary from an ONNX model.

    Parameters
    ----------
    model : ONNX model object.
        Model to be converted to hls model object.
    granularity : string, optional
        How granular you want the configuration to be.
    default_precision : string, optional
        Defines the precsion of your inputs, outputs, weights and biases.
        It is denoted by ap_fixed<X,Y>, where Y is the number of bits representing
        the signed number above the binary point (i.e. the integer part),
        and X is the total number of bits. Additionally, integers in fixed precision
        data type (ap_int<N>, where N is a bit-size from 1 to 1024) can also be used.
    default_reuse_factor : int, optional
        Reuse factor for hls model

    Returns
    -------
    config : dict
        configuration dictionary to be used in ONNX converter.

    See Also
    --------
    hls4ml.config_from_keras_model, hls4ml.convert_from_pytorch_model

    Examples
    --------
    >>> import hls4ml
    >>> config = hls4ml.utils.config_from_keras_model(model, granularity='model')
    >>> hls_model = hls4ml.converters.convert_from_keras_model(model, hls_config=config)

    """

    config = {}

    model_config = {}
    model_config['Precision'] = default_precision
    model_config['ReuseFactor'] = default_reuse_factor
    model_config['Strategy'] = 'Latency'

    config['Model'] = model_config

    return config
